// @sap/cds/lib/srv/protocols/http.js
const cds = require('../../index'), { decodeURI } = cds.utils
const express = require('express')
const PROD = process.env.NODE_ENV === 'production'
const restrict_all = cds.env.requires.auth?.restrict_all_services !== false
const authenticated = PROD && restrict_all ? ['authenticated-user'] : false


class HttpAdapter {

  /** 
   * Constructs and returns a new express.Router 
   * @param {object} srv - The service instance
   * @param {object} [o] - Options for the adapter
   * @returns {express.Router} The constructed router
   */
  constructor (srv,o={}) {
    this.kind = o.kind || this.constructor.name.replace (/Adapter$/,'').toLowerCase()
    this.service = srv
    this.options = o
    return this.router //> constructed by getter
  }


  /** 
   * The actual Router factory. Subclasses override this to add specific handlers. 
   * @returns {express.Router} 
   */
  get router() {
    const router = super.router = new express.Router 
    if (this.requires) router.use (this.requires) //> checks @requires or @restrict roles
    if (this.response4) router.response4 = this.response4     //> used in final error handler
    return router
  }


  /** 
   * Returns a handler to log all incoming requests 
   * @returns {express.RequestHandler}
   */
  get logger() {
    const LOG = cds.log (this.kind)
    return function logger (req,_,next) { 
      LOG.info (req.method, 
        decodeURI (req.baseUrl + req.path),
        Object.keys(req.query).length ? { ...req.query } : ''        
      ); next() 
    }
  }


  /** 
   * Returns a handler to check required roles, or null if no check required. 
   * @returns {express.RequestHandler|null}
   */
  get requires() {
    const d = this.service.definition; if (!d) return null
    const rr = d['@requires'] || d['@restrict']?.map(r => r.to).flat().filter(r => r)
    const roles = !rr?.length ? authenticated : Array.isArray(rr) ? rr : [rr]
    return super.requires = roles && function requires (req, res, next) {
      const user = cds.context.user
      if (roles.some(role => user.has(role))) return next() //> ok
      else if (user._is_anonymous) return next(401) //> request login
      else cds.error (403, `User '${user.id}' is lacking required roles: [${roles}]`)
    }
  }


  /**
   * Returns options for body parsers, e.g., to set a custom limit via 
   * @cds.server.body_parser.limit annotation.
   * @returns {object} Options for body parsers
   */
  get body_parser_options() {
    let options = cds.env.server.body_parser
    let limit = this.service.definition?.['@cds.server.body_parser.limit']
    if (limit) options = { ...options, limit }
    return super.body_parser_options = options
  }


  /**
   * Turn outcomes like results or errors into a proper response body. 
   */
  response4 ({ results, error, errors, messages }) {
    return { data: results, error, errors, messages }
  }
}


module.exports = HttpAdapter
